File: benchmarks/scala/src/main/scala/MtMergeSort.scala
import java.io.{FileInputStream, FileOutputStream}
import java.nio.{ByteBuffer, ByteOrder}
import java.util.concurrent.{Executors, ExecutorService, Future => JFuture}
import scala.concurrent.{Future, Await}
import scala.concurrent.duration._
import scala.concurrent.ExecutionContext.Implicits.global

object MtMergeSort {
  def merge(arr: Array[Int], left: Int, mid: Int, right: Int): Unit = {
    val n1 = mid - left + 1
    val n2 = right - mid

    // Create temporary arrays
    val L = new Array[Int](n1)
    val R = new Array[Int](n2)

    // Copy data to temporary arrays
    System.arraycopy(arr, left, L, 0, n1)
    System.arraycopy(arr, mid + 1, R, 0, n2)

    // Merge the temporary arrays back
    var i = 0
    var j = 0
    var k = left
    while (i < n1 && j < n2) {
      if (L(i) <= R(j)) {
        arr(k) = L(i)
        i += 1
      } else {
        arr(k) = R(j)
        j += 1
      }
      k += 1
    }

    // Copy remaining elements of L[]
    while (i < n1) {
      arr(k) = L(i)
      i += 1
      k += 1
    }

    // Copy remaining elements of R[]
    while (j < n2) {
      arr(k) = R(j)
      j += 1
      k += 1
    }
  }

  def mergeSortThread(arr: Array[Int], left: Int, right: Int, depth: Int, maxDepth: Int, executor: ExecutorService): Unit = {
    if (left < right) {
      val mid = left + (right - left) / 2

      if (depth < maxDepth) {
        // Create futures for left and right halves
        val leftFuture = executor.submit(new Runnable {
          def run(): Unit = mergeSortThread(arr, left, mid, depth + 1, maxDepth, executor)
        })
        val rightFuture = executor.submit(new Runnable {
          def run(): Unit = mergeSortThread(arr, mid + 1, right, depth + 1, maxDepth, executor)
        })

        // Wait for both futures to complete
        leftFuture.get()
        rightFuture.get()
      } else {
        // Sequential sorting for remaining depth
        mergeSortThread(arr, left, mid, depth + 1, maxDepth, executor)
        mergeSortThread(arr, mid + 1, right, depth + 1, maxDepth, executor)
      }

      merge(arr, left, mid, right)
    }
  }

  def main(args: Array[String]): Unit = {
    if (args.length != 4) {
      System.err.println("Usage: scala MtMergeSort <input_file> <num_integers> <num_cores> <output_file>")
      sys.exit(1)
    }

    val inputFile = args(0)
    val numIntegers = args(1).toInt
    val numCores = args(2).toInt
    val outputFile = args(3)

    // Calculate max depth for thread creation
    var maxDepth = 0
    var temp = numCores
    while (temp > 1) {
      maxDepth += 1
      temp /= 2
    }

    // Allocate array for integers
    val arr = new Array[Int](numIntegers)
    val executor = Executors.newFixedThreadPool(numCores)

    try {
      // Read input file
      val fis = new FileInputStream(inputFile)
      val bytes = new Array[Byte](numIntegers * 4)
      val bytesRead = fis.read(bytes)
      fis.close()

      if (bytesRead != numIntegers * 4) {
        System.err.println("Error reading input file")
        sys.exit(1)
      }

      // Convert bytes to integers
      val bb = ByteBuffer.wrap(bytes)
      bb.order(ByteOrder.LITTLE_ENDIAN)
      bb.asIntBuffer().get(arr)

      // Perform parallel merge sort
      mergeSortThread(arr, 0, numIntegers - 1, 0, maxDepth, executor)

      // Write output file
      val fos = new FileOutputStream(outputFile)
      val outBuffer = ByteBuffer.allocate(numIntegers * 4)
      outBuffer.order(ByteOrder.LITTLE_ENDIAN)
      outBuffer.asIntBuffer().put(arr)
      fos.write(outBuffer.array())
      fos.close()
    } catch {
      case e: java.io.FileNotFoundException =>
        System.err.println("Error opening file: " + e.getMessage)
        sys.exit(1)
      case e: Exception =>
        System.err.println("Error: " + e.getMessage)
        sys.exit(1)
    } finally {
      executor.shutdown()
      executor.awaitTermination(1, java.util.concurrent.TimeUnit.MINUTES)
    }
  }
}

File: benchmarks/csharp/src/MtMergeSort.cs
using System;
using System.IO;
using System.Threading.Tasks;

class MtMergeSort
{
    static void Merge(int[] arr, int left, int mid, int right)
    {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        // Create temporary arrays
        int[] L = new int[n1];
        int[] R = new int[n2];

        // Copy data to temporary arrays
        Array.Copy(arr, left, L, 0, n1);
        Array.Copy(arr, mid + 1, R, 0, n2);

        // Merge the temporary arrays back
        int i = 0, j = 0, k = left;
        while (i < n1 && j < n2)
        {
            if (L[i] <= R[j])
            {
                arr[k] = L[i];
                i++;
            }
            else
            {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        // Copy remaining elements of L[]
        while (i < n1)
        {
            arr[k] = L[i];
            i++;
            k++;
        }

        // Copy remaining elements of R[]
        while (j < n2)
        {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    static void MergeSort(int[] arr, int left, int right)
    {
        if (left < right)
        {
            int mid = (left + right) / 2;
            MergeSort(arr, left, mid);
            MergeSort(arr, mid + 1, right);
            Merge(arr, left, mid, right);
        }
    }

    static void ParallelMergeSort(int[] arr, int left, int right, int depth, int maxDepth)
    {
        if (left < right)
        {
            int mid = left + (right - left) / 2;  // Changed to match C version's calculation

            if (depth < maxDepth)
            {
                // Create tasks for parallel execution
                var leftTask = Task.Run(() => ParallelMergeSort(arr, left, mid, depth + 1, maxDepth));
                var rightTask = Task.Run(() => ParallelMergeSort(arr, mid + 1, right, depth + 1, maxDepth));

                // Wait for both tasks to complete
                Task.WaitAll(leftTask, rightTask);
            }
            else
            {
                // Sequential execution for remaining depth
                ParallelMergeSort(arr, left, mid, depth + 1, maxDepth);
                ParallelMergeSort(arr, mid + 1, right, depth + 1, maxDepth);
            }

            Merge(arr, left, mid, right);
        }
    }

    static int Main(string[] args)
    {
        if (args.Length != 4)
        {
            Console.Error.WriteLine("Usage: dotnet run <input_file> <num_integers> <num_cores> <output_file>");
            return 1;
        }

        string inputFile = args[0];
        if (!int.TryParse(args[1], out int numIntegers))
        {
            Console.Error.WriteLine("Invalid number of integers");
            return 1;
        }
        if (!int.TryParse(args[2], out int numCores))
        {
            Console.Error.WriteLine("Invalid number of cores");
            return 1;
        }
        string outputFile = args[3];

        // Calculate max depth based on number of cores (matching C version's approach)
        int maxDepth = 0;
        int temp = numCores;
        while (temp > 1)
        {
            maxDepth++;
            temp /= 2;
        }

        // Allocate array
        int[] arr = new int[numIntegers];

        try
        {
            // Read input file
            using (FileStream fs = new FileStream(inputFile, FileMode.Open, FileAccess.Read))
            {
                byte[] bytes = new byte[numIntegers * sizeof(int)];
                int bytesRead = fs.Read(bytes, 0, bytes.Length);
                if (bytesRead != bytes.Length)
                {
                    Console.Error.WriteLine("Error reading input file");
                    return 1;
                }
                Buffer.BlockCopy(bytes, 0, arr, 0, bytes.Length);
            }

            // Perform parallel merge sort
            ParallelMergeSort(arr, 0, numIntegers - 1, 0, maxDepth);

            // Write output file
            using (FileStream fs = new FileStream(outputFile, FileMode.Create, FileAccess.Write))
            {
                byte[] outputBytes = new byte[numIntegers * sizeof(int)];
                Buffer.BlockCopy(arr, 0, outputBytes, 0, outputBytes.Length);
                fs.Write(outputBytes, 0, outputBytes.Length);
            }
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error: {ex.Message}");
            return 1;
        }

        return 0;
    }
}
